[openxlsx] Add new port

motazmuhammad opened the pull request:
<!-- If your PR fixes issues, please note that here by adding "Fixes #NNNNNN." for each fixed issue on separate lines. -->
Fixes #40931




- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [x] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [x] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html).
- [x] The versioning scheme in `vcpkg.json` matches what upstream says.
- [x] The license declaration in `vcpkg.json` matches what upstream says.
- [x] The installed as the "copyright" file matches what upstream says.
- [x ] The source code of the component installed comes from an authoritative source.
- [x] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is in the new port's versions file.
- [x] Only one version is added to each modified port's versions file.

motazmuhammad committed:
original cmake file

motazmuhammad committed:
Merge branch 'master' of https://github.com/microsoft/vcpkg

motazmuhammad committed:
version database

motazmuhammad committed:
remove unnecessary lines

motazmuhammad committed:
necessary updates

motazmuhammad committed:
add usage file

motazmuhammad committed:
add usage file

motazmuhammad committed:
Merge branch 'microsoft:master' into OpenXLSX

motazmuhammad commented:
> @motazmuhammad please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement

@microsoft-github-policy-service agree

motazmuhammad committed:
Merge branch 'microsoft:master' into OpenXLSX

MonicaLiu0311 commented:
Why add CMakeLists.txt for libxml? It's already provided [upstream](https://gitlab.gnome.org/GNOME/libxml2/-/blob/v2.11.9/CMakeLists.txt?ref_type=tags).

MonicaLiu0311 commented:
New port
1. You can create the yalantinglibs folder in `vcpkg/ports/` and add `portfile.cmake` and `vcpkg.json`:
```
vcpkg/ports/testport/
    portfile.cmake
    vcpkg.json
```
Then run `./vcpkg install testport`, if the installation is successful, it proves that there is no problem with the files you added.

2. Next, run `./vcpkg x-add-verison testport`, the `vcpkg/verisons/t-/testport.josn` file will be added automatically, and a record will be added in `vcpkg/verisons/baseline.json`.
```
vcpkg/versions/
    t-/
        testport.josn (automatically generated)
    baseline.json (automatically add records)
```

The modification needs to meet the requirements: [maintainer-guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)

MonicaLiu0311 commented on ports/openxlsx/vcpkg.json:
```diff
@@ -0,0 +1,17 @@
+{
+  "name": "openxlsx",
+  "version": "1.4.0",
```
`"version"` must be consistent with upstream. I found that the latest version is `0.3.2`: https://github.com/troldal/OpenXLSX/tags.

motazmuhammad replied:
While the latest stable (tagged) is 0.3.2 the latest is 0.4.1 This can be confirmed from the cmake file 
https://github.com/troldal/OpenXLSX/blob/master/CMakeLists.txt

I chose to opt for using0.4.1 to be consistent with conan which uses 0.4.1 as shown here
https://conan.io/center/recipes/openxlsx
Please be patient as I dont fully understand the vcpkg maintainer guide(I tried to follow it to the best of my understanding).

There is still a mistake, however, in my commit as it should be 0.4.1 instead of 1.4.1

dg0yt replied:
OpenXSLX must define which git ref is the release. Usually this is done with a tag.
Conan is not a suitable reference. CMakeLists.txt is not a precise reference. (The "new development branch" uses the same version number.)
Until the projects make proper releases (again), the port might use `version-date`.

MonicaLiu0311 commented on ports/openxlsx/usage:
```diff
@@ -0,0 +1,2 @@
+  find_package(OpenXLSX CONFIG REQUIRED)
+  target_link_libraries(main PRIVATE OpenXLSX::OpenXLSX)
```
```suggestion
OpenXLSX provides CMake targets:

    find_package(OpenXLSX CONFIG REQUIRED)
    target_link_libraries(main PRIVATE OpenXLSX::OpenXLSX)

```

MonicaLiu0311 replied:
The last empty line is a format requirement.

motazmuhammad replied:
Thank you for clarifying that to me

dg0yt replied:
Follow the pattern from vcpkg tool.
- The original indentation was right.
- There is confusion about the end of the text. What is not acceptable, and what is indicated on GH by a red `(-)`, is the text ending without a line break. In VS Code it means to add a new empty line. In most UNIX text editors, lines are always ended with a line break.

```suggestion
openxlsx provides CMake targets:

  find_package(OpenXLSX CONFIG REQUIRED)
  target_link_libraries(main PRIVATE OpenXLSX::OpenXLSX)
```

MonicaLiu0311 commented on ports/openxlsx/portfile.cmake:
```diff
@@ -0,0 +1,23 @@
+    SHA512 e3d27b6e21dcb727398ec1c78cdde4adfb9d5d354380e22217f63c46dc63a1dcdfde54733b70ac9b7e5fa7f6bc47028b311f03fdccc980ed9132ac0a23a23ed4
+    HEAD_REF master
+)
+vcpkg_configure_cmake(
```
Please follow the prompts to modify the deprecated functions.

MonicaLiu0311 commented:
> New port
> 
> 1. You can create the yalantinglibs folder in `vcpkg/ports/` and add `portfile.cmake` and `vcpkg.json`:
> 
> ```
> vcpkg/ports/testport/
>     portfile.cmake
>     vcpkg.json
> ```
> 
> Then run `./vcpkg install testport`, if the installation is successful, it proves that there is no problem with the files you added.
> 
> 2. Next, run `./vcpkg x-add-verison testport`, the `vcpkg/verisons/t-/testport.josn` file will be added automatically, and a record will be added in `vcpkg/verisons/baseline.json`.
> 
> ```
> vcpkg/versions/
>     t-/
>         testport.josn (automatically generated)
>     baseline.json (automatically add records)
> ```
> 
> The modification needs to meet the requirements: [maintainer-guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)

Reminder: The version file is missing. Please refer to: [39493/files](https://github.com/microsoft/vcpkg/pull/39493/files).

motazmuhammad committed:
Revert "original cmake file"

This reverts commit ea6126637792e76ab9125a98788c121c334f61db.

motazmuhammad committed:
Merge branch 'OpenXLSX' of https://github.com/motazmuhammad/vcpkg into OpenXLSX

motazmuhammad committed:
respond to comments

motazmuhammad committed:
add a version file and update the baseline.json

motazmuhammad committed:
change to 0.3.2

motazmuhammad committed:
add the versions changes and baseline

motazmuhammad commented:
@MonicaLiu0311 Thank you very much for all your insightful comments, I really appreciate them, I updated the fork with your suggested changes ( according to my best understanding). Please let me know if you have further comments, and please be patient with me).


Thank you again

MonicaLiu0311 commented:
When testing usage, the following error occurs:
```
E:\test\build>cmake --build .
MSBuild version 17.11.2+c078802d4 for .NET Framework

  1>Checking Build System
  Building Custom Rule E:/test/CMakeLists.txt
  test.cpp
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.41.34120\include\variant(12): warning STL4038
: The contents of <variant> are available only with C++17 or later. [E:\test\build\main.vcxproj]
E:\openxlsx\installed\x64-windows\include\OpenXLSX\headers\XLXmlParser.hpp(50,10): error C1083: Cannot open include fil
e: 'external/pugixml/pugixml.hpp': No such file or directory [E:\test\build\main.vcxproj]
  (compiling source file '../test.cpp')
```

<details>
<summary><b>test.cpp</b></summary>
<pre>
#include &#60;iostream&#62;
#include "OpenXLSX.hpp"

using namespace std;

int main()
{
        cout << "Hello CMake." << endl;
        return 0;
}
</pre></details>


<details>
<summary><b>CMakeLists.txt</b></summary>
<pre>
cmake_minimum_required (VERSION 3.8)

set(CMAKE_TOOLCHAIN_FILE "E:/openxlsx/scripts/buildsystems/vcpkg.cmake")

project ("test")

add_library (main "test.cpp")

find_package(OpenXLSX CONFIG REQUIRED)
target_link_libraries(main PRIVATE OpenXLSX::OpenXLSX)
</pre></details>

motazmuhammad committed:
fix missing dependencies

motazmuhammad commented:
> When testing usage, the following error occurs:
> 
> ```
> E:\test\build>cmake --build .
> MSBuild version 17.11.2+c078802d4 for .NET Framework
> 
>   1>Checking Build System
>   Building Custom Rule E:/test/CMakeLists.txt
>   test.cpp
> C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.41.34120\include\variant(12): warning STL4038
> : The contents of <variant> are available only with C++17 or later. [E:\test\build\main.vcxproj]
> E:\openxlsx\installed\x64-windows\include\OpenXLSX\headers\XLXmlParser.hpp(50,10): error C1083: Cannot open include fil
> e: 'external/pugixml/pugixml.hpp': No such file or directory [E:\test\build\main.vcxproj]
>   (compiling source file '../test.cpp')
> ```
> 
> test.cpp
> CMakeLists.txt

@MonicaLiu0311 Thank you very much that is fixed now. I have a question, if you dont mind please reply to them.

1. Right now I  am using the head revision (master branch) as evident by the portfile     REF master
is that ok, I tried using     REF "${VERSION}" however, the correct file was not picked up unless i make it 
    REF "${VERSION}"
    REF "refs/tags/v${VERSION}"
and if I do that, I need to change the SHA512 , do I need to create a seperate SHA for each version? I could not figure this out from the documents.

Thank you again for your patience. I hope to become a regular contributor to vcpkg

MonicaLiu0311 commented on ports/openxlsx/portfile.cmake:
```diff
@@ -0,0 +1,27 @@
+vcpkg_from_github(
+    OUT_SOURCE_PATH SOURCE_PATH
+    REPO troldal/OpenXLSX
+    REF master
+    SHA512 429215a961cfcfe8fcb49ee48970dd10dccad09a5c80db2eea2413120a51a2529c9a650584a079756c3b463d53f1fcdec46fdc158ffbc680592896867f60d3ea
```
```suggestion
    REF "v${VERSION}"
    SHA512 52205b394383d45c0fb16599ab96453d8a5b9b5cd596096848cc888f47565b2713d9edded06b2ecd7b67736622badf136e4b1becc57bfa5bbdcb1e063a347084
```

MonicaLiu0311 commented:
>    REF "refs/tags/v${VERSION}"

The regular expression for `${VERSION}` matches only [VERSION semantics](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/users/versioning.md).

motazmuhammad committed:
Add a batch file and use the version string
Notes:
the patch file is to fix a compilation error in vs2019, 2022, this was already fixed in the master but not committed to the latest release

motazmuhammad committed:
add the patch file

motazmuhammad committed:
fix the version file

motazmuhammad commented:
> When testing usage, the following error occurs:
> 
> ```
> E:\test\build>cmake --build .
> MSBuild version 17.11.2+c078802d4 for .NET Framework
> 
>   1>Checking Build System
>   Building Custom Rule E:/test/CMakeLists.txt
>   test.cpp
> C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.41.34120\include\variant(12): warning STL4038
> : The contents of <variant> are available only with C++17 or later. [E:\test\build\main.vcxproj]
> E:\openxlsx\installed\x64-windows\include\OpenXLSX\headers\XLXmlParser.hpp(50,10): error C1083: Cannot open include fil
> e: 'external/pugixml/pugixml.hpp': No such file or directory [E:\test\build\main.vcxproj]
>   (compiling source file '../test.cpp')
> ```
> 
> test.cpp
> CMakeLists.txt

This now works, but please note that to be able to use openxlsx in the cmakelists.txt file one has to add the following line set(CMAKE_CXX_STANDARD 17) as the some of the header files are dependent on c++17, should that be mentioned in the usage file in one way or the other?

motazmuhammad committed:
Merge branch 'microsoft:master' into OpenXLSX

dg0yt commented:
> 1. Right now I  am using the head revision (master branch) as evident by the portfile
> ~~~
>     REF master
> ~~~
>    is that ok, 

No. `REF` must identify a particular immutable release, tag or commit. `master` is a branch name, and the contents may change. (`HEAD_REF` may point to `master`.)

> I tried using
> ~~~
>    REF "${VERSION}" 
> ~~~
> however, the correct file was not picked up unless i make it
> ~~~
>   REF "refs/tags/v${VERSION}"
> ~~~
This indicates that the tags are named `v<version>`, and you should use
~~~
   REF "v${VERSION}"
~~~
> and if I do that, I need to change the SHA512 , do I need to create a seperate SHA for each version?

Yes. The SHA512 is to verify the downloaded archive, and it is also used an source asset cache key.


> note that to be able to use openxlsx in the cmakelists.txt file one has to add the following line set(CMAKE_CXX_STANDARD 17) as the some of the header files are dependent on c++17, should that be mentioned in the usage file in one way or the other?

If there is exported CMake config, the C++17 requirement could be in the exported compile features.
https://cmake.org/cmake/help/latest/manual/cmake-compile-features.7.html#requiring-language-standards

dg0yt commented on ports/openxlsx/compilation_fix.patch:
```diff
@@ -0,0 +1,119 @@
+          * @return
+          * @todo To be implemented.
+          */
+-        XLColor color() const
+-        {
+-            return XLColor();
+-        }
++        XLColor color() const { return XLColor(); }
```
This patch has several such sections which are not needed to fix problems.
Either let the portfile download the original raw patch from upstream's repo,
or create a minimal patch of necessary changes.

dg0yt commented on ports/openxlsx/portfile.cmake:
```diff
@@ -0,0 +1,31 @@
+    SHA512 52205b394383d45c0fb16599ab96453d8a5b9b5cd596096848cc888f47565b2713d9edded06b2ecd7b67736622badf136e4b1becc57bfa5bbdcb1e063a347084
+    HEAD_REF master
+    PATCHES
+    compilation_fix.patch
```
Nit-pick:
```suggestion
        compilation_fix.patch
```

dg0yt commented on ports/openxlsx/portfile.cmake:
```diff
@@ -0,0 +1,31 @@
+    PATCHES
+    compilation_fix.patch
+)
+set(CMAKE_CXX_STANDARD 17)
+
+vcpkg_cmake_configure(
+    SOURCE_PATH ${SOURCE_PATH}
+    PREFER_NINJA
+    OPTIONS
+        -DOPENXLSX_BUILD_TESTS=OFF
+)
```
```suggestion
vcpkg_cmake_configure(
    SOURCE_PATH ${SOURCE_PATH}
    PREFER_NINJA
    OPTIONS
        -DCMAKE_CXX_STANDARD=17
        -DOPENXLSX_BUILD_TESTS=OFF
)
```
Portfiles run in script mode. `CMAKE_CXX_STANDARD` has no effect.
https://learn.microsoft.com/en-us/vcpkg/contributing/maintainer-guide#portfiles-are-run-in-script-mode

dg0yt commented on ports/openxlsx/portfile.cmake:
```diff
@@ -0,0 +1,31 @@
+
+vcpkg_cmake_configure(
+    SOURCE_PATH ${SOURCE_PATH}
+    PREFER_NINJA
```
```suggestion
```
Default.

dg0yt commented on ports/openxlsx/portfile.cmake:
```diff
@@ -0,0 +1,31 @@
+    OPTIONS
+        -DOPENXLSX_BUILD_TESTS=OFF
+)
+vcpkg_cmake_install(DISABLE_PARALLEL)
```
Why `DISABLE_PARALLEL`?

motazmuhammad replied:
That is not needed, nice catch, thank you very much

dg0yt commented on ports/openxlsx/usage:
```diff
@@ -0,0 +1,3 @@
+  find_package(OpenXLSX CONFIG REQUIRED)
+  target_link_libraries(main PRIVATE OpenXLSX::OpenXLSX)
+  
```
No red `(-)` at the end, please.
```suggestion
openxlsx provides CMake targets:

  find_package(OpenXLSX CONFIG REQUIRED)
  target_link_libraries(main PRIVATE OpenXLSX::OpenXLSX)
```

motazmuhammad committed:
reply to comments

motazmuhammad committed:
Merge branch 'OpenXLSX' of https://github.com/motazmuhammad/vcpkg into OpenXLSX

motazmuhammad committed:
fix the version file

motazmuhammad committed:
fix the sha-512 and version file

motazmuhammad committed:
fix the version file

motazmuhammad commented:
@dg0yt thank you very much for your comments, I really appreciate all of them, I have replied to them to the best of my knowledge. Please accept my apologies for taking too much of your time, this is my first contribution to vcpkg

motazmuhammad committed:
Update ports/openxlsx/usage

Co-authored-by: Kai Pastor <dg0yt@darc.de>

motazmuhammad committed:
Update usage

motazmuhammad committed:
fix version file

motazmuhammad commented:
@MonicaLiu0311 @dg0yt is there anything else I need to do?

motazmuhammad committed:
Merge branch 'microsoft:master' into OpenXLSX

MonicaLiu0311 commented:
The usage test passed on `x64-windows` (header files found):
```
openxlsx provides CMake targets:

  find_package(OpenXLSX CONFIG REQUIRED)
  target_link_libraries(main PRIVATE OpenXLSX::OpenXLSX)
```

JavierMatosD commented on ports/openxlsx/portfile.cmake:
```diff
@@ -0,0 +1,30 @@
+# Header-only library
```
```suggestion
# Header-only library
set(VCPKG_BUILD_TYPE release)
```

JavierMatosD commented on ports/openxlsx/portfile.cmake:
```diff
@@ -0,0 +1,30 @@
+vcpkg_cmake_configure(
+    SOURCE_PATH ${SOURCE_PATH}
+    OPTIONS
+        -DCMAKE_CXX_STANDARD=17
```
```suggestion
```

This has no effect.

JavierMatosD commented on ports/openxlsx/portfile.cmake:
```diff
@@ -0,0 +1,30 @@
+vcpkg_cmake_install()
+vcpkg_cmake_config_fixup(CONFIG_PATH lib/cmake/OpenXLSX)
+file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/include")
+file(COPY "${SOURCE_PATH}/OpenXLSX/external" DESTINATION "${CURRENT_PACKAGES_DIR}/include")
```
External? It looks like this is vending `zippy`, `pugixml`, and `nowide`

Pleas see https://learn.microsoft.com/en-us/vcpkg/contributing/maintainer-guide#do-not-use-vendored-dependencies

motazmuhammad replied:
@JavierMatosD one of the dependencies (zippy) is not in vcpkg, should it be added to vcpkg, in a separate port as a requirement for this port?

JavierMatosD replied:
Yes. Please open another PR adding `zippy` as a port as it is a requirement for this port.

JavierMatosD commented on ports/openxlsx/portfile.cmake:
```diff
@@ -0,0 +1,30 @@
+    SOURCE_PATH ${SOURCE_PATH}
+    OPTIONS
+        -DCMAKE_CXX_STANDARD=17
+        -DOPENXLSX_BUILD_TESTS=OFF
```
```suggestion
        -DOPENXLSX_BUILD_TESTS=OFF
        -OPENXLSX_CREATE_DOCS=OFF
        -OPENXLSX_BUILD_SAMPLES=OFF
        -DOPENXLSX_BUILD_BENCHMARKS=OFF
```

JavierMatosD commented on ports/openxlsx/compilation_fix.patch:
```diff
@@ -0,0 +1,22 @@
+         {
+             XLQuery query(XLQueryType::QuerySheetVisibility);
+-            query.template setParam("sheetID", relationshipID());
++            query.setParam("sheetID", relationshipID());
```
Why is this patch needed? Has it been reported upstream?

motazmuhammad replied:
@JavierMatosD Yes, it has been fixed in the upstream but it has not been put in the release refs yet, it is in the master (main ) branch, however.

JavierMatosD replied:
If this patch is already fixed upstream, can you download and apply the patch directly instead of checking it in?

JavierMatosD reviewed (commented):
Partial Review:

It looks like this port is vendoring dependencies. All the ports dependencies should be explicitly listed in the dependency section of the `vcpkg.json`.

motazmuhammad committed:
Update ports/openxlsx/portfile.cmake

Co-authored-by: Javier Matos Denizac <javier.matosd@gmail.com>

motazmuhammad committed:
Update ports/openxlsx/portfile.cmake

Co-authored-by: Javier Matos Denizac <javier.matosd@gmail.com>

MonicaLiu0311 commented:
Is there any new progress?

motazmuhammad commented:
@MonicaLiu0311 I apologize, I got really busy with personal stuff. I will be back to the library in less than a week from now.

MonicaLiu0311 commented:
Ping @motazmuhammad

MonicaLiu0311 commented:
Closing this PR since it seems that no progress is being made. Please ping us to reopen if work is still being done.
